File: easyremote/_host/forward.c
/*
 * easyremote fast forwarder — the low-latency interim path (SPEC D2).
 *
 * stdin(args JSON) -> warm-host Unix socket -> stdout(result JSON).
 * Replaces the Python shim's ~50-60ms interpreter spawn with a
 * single-digit-millisecond native spawn; removed entirely once the
 * daemon host-attach protocol (Cli PR-1) lands.
 *
 * Response slicing relies on the host's PINNED serialization prefixes
 * (tests/test_host.py::test_wire_prefixes_are_pinned_for_the_c_forwarder):
 *   {"ok":true,"result":<json>}   ->  print <json>, exit 0
 *   {"ok":false,"error":<json>}   ->  print <json> to stderr, exit 1
 *
 * Build: cc -O2 -o easyremote-forward forward.c   (done lazily by
 * easyremote._host.fastpath; Python shim remains the fallback).
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <unistd.h>

static int fail(const char *msg) {
    fprintf(stderr, "%s\n", msg);
    return 1;
}

static char *read_all(FILE *stream, size_t *out_len) {
    size_t cap = 1 << 16, len = 0, n;
    char *buf = malloc(cap);
    if (!buf) return NULL;
    while ((n = fread(buf + len, 1, cap - len - 1, stream)) > 0) {
        len += n;
        if (len + 1 >= cap) {
            char *grown = realloc(buf, cap *= 2);
            if (!grown) { free(buf); return NULL; }
            buf = grown;
        }
    }
    buf[len] = 0;
    *out_len = len;
    return buf;
}

int main(int argc, char **argv) {
    if (argc != 3)
        return fail("usage: easyremote-forward <socket> <fn>  (args on stdin)");
    const char *path = argv[1], *fn = argv[2];

    size_t in_len = 0;
    char *in = read_all(stdin, &in_len);
    if (!in) return fail("out of memory");
    while (in_len && (unsigned char)in[in_len - 1] <= ' ') in[--in_len] = 0;
    const char *args = in_len ? in : "{}";

    int sock = socket(AF_UNIX, SOCK_STREAM, 0);
    if (sock < 0) return fail("socket() failed");
    struct sockaddr_un addr;
    memset(&addr, 0, sizeof addr);
    addr.sun_family = AF_UNIX;
    if (strlen(path) >= sizeof addr.sun_path) return fail("socket path too long");
    strcpy(addr.sun_path, path);
    if (connect(sock, (struct sockaddr *)&addr, sizeof addr) != 0) {
        fprintf(stderr,
                "easyremote host unreachable at %s — is the ComputeNode process"
                " running?\n", path);
        return 1;
    }

    /* fn names are validated upstream to [A-Za-z0-9_.-]: no escaping needed. */
    size_t req_cap = strlen(fn) + strlen(args) + 32;
    char *req = malloc(req_cap);
    if (!req) return fail("out of memory");
    int req_len = snprintf(req, req_cap, "{\"fn\":\"%s\",\"args\":%s}\n", fn, args);
    if (send(sock, req, (size_t)req_len, 0) != req_len) return fail("send failed");

    size_t cap = 1 << 16, len = 0;
    char *resp = malloc(cap);
    if (!resp) return fail("out of memory");
    ssize_t got;
    while ((got = recv(sock, resp + len, cap - len - 1, 0)) > 0) {
        len += (size_t)got;
        if (resp[len - 1] == '\n') break;
        if (len + 1 >= cap) {
            char *grown = realloc(resp, cap *= 2);
            if (!grown) return fail("out of memory");
            resp = grown;
        }
    }
    while (len && (resp[len - 1] == '\n' || resp[len - 1] == '\r')) len--;
    resp[len] = 0;

    static const char OK[] = "{\"ok\":true,\"result\":";
    static const char ERR[] = "{\"ok\":false,\"error\":";
    if (len > sizeof OK - 1 && memcmp(resp, OK, sizeof OK - 1) == 0) {
        fwrite(resp + sizeof OK - 1, 1, len - (sizeof OK - 1) - 1, stdout);
        return 0;
    }
    if (len > sizeof ERR - 1 && memcmp(resp, ERR, sizeof ERR - 1) == 0) {
        fprintf(stderr, "%.*s\n", (int)(len - (sizeof ERR - 1) - 1),
                resp + sizeof ERR - 1);
        return 1;
    }
    return fail("easyremote host sent an unrecognized response");
}

File: easyremote/_host/forward.py
"""Per-invocation forwarder: the stdin/stdout shim between daemon and warm host.

The device-ability runtime invokes ``command`` with the args JSON on
stdin and reads the result JSON from stdout (contract documented in the
``easynet ability new`` scaffold). This shim relays stdin to the
resident :class:`~easyremote._host.HostServer` over its Unix socket and
prints the host's result — JSON types arrive intact, no re-typing
anywhere. Exit 0 with the result on stdout; exit 1 with the error on
stderr (surfaced to the caller by the runtime).

This shim is the transition cost of SPEC §4.1 D2: one interpreter
spawn per call until the daemon host-attach protocol (Cli PR-1)
removes it.
"""

from __future__ import annotations

import json
import socket
import sys


def main(argv: list[str] | None = None) -> int:
    argv = sys.argv[1:] if argv is None else argv
    if len(argv) != 2:
        print(
            "usage: python -m easyremote._host.forward <socket> <fn>  (args on stdin)",
            file=sys.stderr,
        )
        return 1
    socket_path, fn_name = argv

    stdin_payload = sys.stdin.read().strip()
    try:
        args = json.loads(stdin_payload) if stdin_payload else {}
    except json.JSONDecodeError as exc:
        print(f"stdin is not valid JSON: {exc}", file=sys.stderr)
        return 1

    request = json.dumps({"fn": fn_name, "args": args}, separators=(",", ":"))
    try:
        with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as connection:
            connection.connect(socket_path)
            connection.sendall(request.encode("utf-8") + b"\n")
            response = _read_line(connection)
    except OSError as exc:
        print(
            f"easyremote host unreachable at {socket_path}: {exc}"
            " — is the ComputeNode process running?",
            file=sys.stderr,
        )
        return 1

    try:
        payload = json.loads(response)
    except json.JSONDecodeError as exc:
        print(f"easyremote host sent invalid JSON: {exc}", file=sys.stderr)
        return 1

    if payload.get("ok"):
        sys.stdout.write(json.dumps(payload.get("result"), separators=(",", ":")))
        return 0
    error = payload.get("error", {})
    print(
        f"{error.get('kind', 'INTERNAL')}/{error.get('reason', '')}:"
        f" {error.get('message', '')}",
        file=sys.stderr,
    )
    return 1


def _read_line(connection: socket.socket) -> str:
    chunks: list[bytes] = []
    while True:
        chunk = connection.recv(65536)
        if not chunk:
            break
        chunks.append(chunk)
        if chunk.endswith(b"\n"):
            break
    return b"".join(chunks).decode("utf-8")


if __name__ == "__main__":
    raise SystemExit(main())
